pages/_app/index.jsx

```diff
@@ -35,19 +32,24 @@ class App extends NextApp {
 
     event.preventDefault()
     const route = new Route(url.pathname)
-    this.props.router.push(route.href, route.as)
+    await this.props.router.push(route.href, route.as)
   }
 
   render() {
-    const { Component, pageProps, initialStoreData, router } = this.props
+    const { Component, pageProps, store, router } = this.props
 
     const { dataProvider, activity } = new Route(router.asPath)
 
+    // needs to be done here because getInitialProps can be called before
+    // handleNavigation (router.push) is done
+    store.dataProvider = dataProvider
+    store.activity = activity
+
     return (
-      <GlobalProvider initialData={initialStoreData}>
+      <GlobalProvider store={store}>
         <Application
-          dataProvider={dataProvider}
-          activity={activity}
+          dataProvider={store.dataProvider}
```

Comment:
I think, it needs only `dataProviderId`, that can be passed as `dataProvider.id`. Everything inside can be connected independently.

I try to pass primitives to props as much as possible because passing object to HTML attribute seems to be weird. It's only because of JSX notation, I feel good passing objects using `React.createElement(Component, props, children)`

Reply:
Ok let me change it in next PR